compiler/rustc_typeck/src/check/expr.rs

```diff
@@ -2347,16 +2347,63 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 suggested_field_name.to_string(),
                 Applicability::MaybeIncorrect,
             );
+            return;
+        }
+
+        let struct_variant_def = def.non_enum_variant();
+        let field_names = self.available_field_names(struct_variant_def, access_span);
+        let name = field.name.as_str();
+
+        fn get_common_prefix_len(a: &str, b: &str) -> usize {
+            let common_bytes_count =
+                a.as_bytes().iter().zip(b.as_bytes().iter()).take_while(|(c, d)| c == d).count();
+
+            return a.floor_char_boundary(common_bytes_count);
+        }
+
+        // Calculate the field programmer most likely wanted to use.
+        let longest_prefix_len = field_names
+            .iter()
+            .map(Symbol::as_str)
+            .map(move |s| get_common_prefix_len(s, name))
+            .max()
+            .unwrap_or(0);
+
+        // Field name will have at least one character.
+        let last_letter = name.chars().last().unwrap();
+        // When the keyboard gets hit, first there will be a few keystrokes registered,
+        // then one button will be "chosen" as continuation.
+        let gibberish = name[longest_prefix_len..].trim_end_matches(last_letter);
+        let letters_asleep = name.len() - longest_prefix_len - gibberish.len();
+
+        use std::time::Duration;
+        const KEY_REPEAT_PER_SEC: u64 = 25;
+        const KEY_REPEAT_DELAY: Duration = Duration::from_millis(600);
+        const TIME_PER_KEY_REPEAT: Duration = Duration::from_millis(1000 / KEY_REPEAT_PER_SEC);
+
+        let asleep_for = KEY_REPEAT_DELAY + TIME_PER_KEY_REPEAT * letters_asleep as u32;
+        if asleep_for > Duration::from_secs(1) {
```

Comment:
I think it would be clearer if this checked for the character amount directly, to make it clear how many are required for the hint.